cloud/cluster/cluster.go

```diff
@@ -1,51 +1,121 @@
 package cluster
 
 import (
-	"fmt"
-	"io"
+	"sort"
 )
 
-// NodeUpdate represents a change to the cluster
-type NodeUpdate struct {
-	UpdateType NodeUpdateType
-	Id         NodeId
-	Node       Node // Only set for adds
+// Cluster represents a cluster of Nodes.
+
+type Cluster struct {
+	State    *State
+	reqCh    chan interface{}
+	updateCh chan []NodeUpdate
+	StateCh  chan []Node
+	subs     []chan []NodeUpdate
 }
 
-func (u *NodeUpdate) String() string {
-	return fmt.Sprintf("%v %v %v", u.UpdateType, u.Id, u.Node)
+func NewCluster(state []Node, updateCh chan []NodeUpdate, stateCh chan []Node) *Cluster {
+	s := MakeState(state)
+	c := &Cluster{
+		State:    s,
+		reqCh:    make(chan interface{}),
+		updateCh: updateCh,
+		StateCh:  stateCh,
+		subs:     nil,
+	}
+	go c.loop()
+	return c
 }
 
-// Cluster represents a cluster of Nodes.
-type Cluster interface {
-	// Members returns the current members, or an error if they can't be determined.
-	Members() []Node
-	// Subscribe subscribes to changes to the cluster.
-	Subscribe() Subscription
-	// Stop monitoring this cluster
-	Close() error
+func (c *Cluster) Members() []Node {
+	ch := make(chan []Node)
+	c.reqCh <- ch
+	return <-ch
 }
 
-// Subscription is a subscription to cluster changes.
-type Subscription struct {
-	InitialMembers []Node            // The members at the time the subscription started
-	Updates        chan []NodeUpdate // Updates as they happen
-	Closer         io.Closer         // How to stop subscribing
+func (c *Cluster) Subscribe() Subscriber {
+	ch := make(chan Subscriber)
+	c.reqCh <- ch
+	return <-ch
 }
 
-// Helper functions to create NodeUpdates
+func (c *Cluster) Close() error {
```

Comment:
We might want to close all all the subs and clear c.subs, otherwise we're left with subs that can unsubscribe. Just a thought, since we don't have a clear shutdown story afaik.